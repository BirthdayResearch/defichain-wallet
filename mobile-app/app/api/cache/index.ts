import Cache from 'lru-cache'

class LruCache {
  private readonly cache

  constructor () {
    // setting maxAge to 10 min
    this.cache = new Cache({
      max: 1000,
      maxAge: 1000 * 60 * 5
    })
  }

  /**
   * @param key {string} of item with 'network' prefixed
   * @return {string | null}
   */
  get (key: string): any {
    return this.cache.get(key)
  }

  /**
   * @param key {string} of item with 'network' prefixed
   * @param value {any} to set
   */
  set (key: string, value: any, maxAge?: number): void {
    this.cache.set(key, value, maxAge)
  }

  clear (): void {
    this.cache.reset()
  }
}

export const CacheApi = new LruCache()
