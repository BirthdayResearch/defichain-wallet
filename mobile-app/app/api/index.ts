export * from './secured'
export * from './logging'
export { ThemePersistence } from './persistence/theme_storage'
export { LanguagePersistence } from './persistence/language_storage'
export { DisplayDexGuidelinesPersistence } from './persistence/display_dexguidelines_storage'
export { DisplayBalancesPersistence } from './persistence/display_balances_storage'
export { FeatureFlagPersistence } from './persistence/feature_flag_storage'
export { SlippageTolerancePersistence } from './persistence/slippage-tolerance_storage'
