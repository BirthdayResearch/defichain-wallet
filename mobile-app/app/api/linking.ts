import { Platform } from 'react-native'
import * as Linking from 'expo-linking'
import * as WebBrowser from 'expo-web-browser'

export async function openURL (url: string): Promise<void> {
  if (Platform.OS === 'web') {
    window.open(url, '_target')
    return
  }

  if (Platform.OS !== 'android' && await Linking.canOpenURL(url)) {
    await Linking.openURL(url)
    return
  }

  await WebBrowser.openBrowserAsync(url)
}
