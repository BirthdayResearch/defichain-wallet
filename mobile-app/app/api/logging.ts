/**
 * Centralize logging/error reporting for log abstraction
 */
export const Logging = {
  error (error: any): void {
    console.error(error)
  },
  info (message: string): void {
    console.log(message)
  }
}
