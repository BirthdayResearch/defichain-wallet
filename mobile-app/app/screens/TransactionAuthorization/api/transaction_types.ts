export const MAX_PASSCODE_ATTEMPT = 3 // allowed 2 failures
export const PASSCODE_LENGTH = 6
export const INVALID_HASH = 'invalid hash'
export const USER_CANCELED = 'USER_CANCELED'

export const DEFAULT_MESSAGES = {
  message: 'Enter passcode to continue',
  loadingMessage: 'Signing your transaction...',
  authorizedTransactionMessage: {
    title: 'Transaction authorized',
    description: 'Please wait while your transaction is being prepared'
  },
  grantedAccessMessage: {
    title: 'Access granted',
    description: 'You may now proceed'
  }
}

export const SUCCESS_DISPLAY_TIMEOUT_IN_MS = 2000
export const CANCELED_ERROR = 'canceled error'

export enum TransactionStatus {
  INIT,
  IDLE,
  BLOCK,
  PIN,
  SIGNING,
  AUTHORIZED
}

export interface PromptPromiseI {
  resolve: (pin: string) => void
  reject: (e: Error) => void
}
