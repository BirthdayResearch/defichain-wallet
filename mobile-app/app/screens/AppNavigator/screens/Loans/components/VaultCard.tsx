
import BigNumber from 'bignumber.js'
import { ThemedIcon, ThemedText, ThemedTouchableOpacity, ThemedView } from '@components/themed'
import { tailwind } from '@tailwind'
import { View } from '@components'
import { translate } from '@translations'
import { TokenIconGroup } from '@components/TokenIconGroup'
import { IconButton } from '@components/IconButton'
import { NavigationProp, useNavigation } from '@react-navigation/native'
import { LoanParamList } from '@screens/AppNavigator/screens/Loans/LoansNavigator'
import { LoanVault } from '@store/loans'
import { LoanVaultActive, LoanVaultState } from '@defichain/whale-api-client/dist/api/loan'
import { VaultInfo } from '@screens/AppNavigator/screens/Loans/components/VaultInfo'
import { TouchableOpacity } from 'react-native'
import { openURL } from '@api/linking'
import { useDeFiScanContext } from '@shared-contexts/DeFiScanContext'
import { VaultSectionTextRow } from '@screens/AppNavigator/screens/Loans/components/VaultSectionTextRow'
import {
  useVaultStatus,
  VaultStatusTag
} from '@screens/AppNavigator/screens/Loans/components/VaultStatusTag'
import { useNextCollateralizationRatio } from '@screens/AppNavigator/screens/Loans/hooks/NextCollateralizationRatio'
import {
  CollateralizationRatioDisplay
} from '@screens/AppNavigator/screens/Loans/components/CollateralizationRatioDisplay'
import { TabKey } from '@screens/AppNavigator/screens/Loans/VaultDetail/components/VaultDetailTabSection'
import { useLoanOperations } from '@screens/AppNavigator/screens/Loans/hooks/LoanOperations'
import { VaultStatus } from '@screens/AppNavigator/screens/Loans/VaultStatusTypes'
import { getUSDPrecisedPrice } from '@screens/AppNavigator/screens/Auctions/helpers/usd-precision'

export interface VaultCardProps extends React.ComponentProps<any> {
  vault: LoanVault
  testID: string
}

export function VaultCard (props: VaultCardProps): JSX.Element {
  const navigation = useNavigation<NavigationProp<LoanParamList>>()
  const vault = props.vault as LoanVaultActive
  const { getVaultsUrl } = useDeFiScanContext()
  const vaultState = useVaultStatus(vault.state, new BigNumber(vault.collateralRatio), new BigNumber(vault.loanScheme.minColRatio), new BigNumber(vault.loanValue), new BigNumber(vault.collateralValue))
  const nextCollateralizationRatio = useNextCollateralizationRatio(vault.collateralAmounts, vault.loanAmounts)
  const canUseOperations = useLoanOperations(vault?.state)
  const onCardPress = (): void => {
    navigation.navigate('VaultDetailScreen', {
      vaultId: vault.vaultId
    })
  }
  return (
    <ThemedView
      light={tailwind('bg-white border-gray-200')}
      dark={tailwind('bg-gray-800 border-gray-700')}
      style={tailwind('rounded mb-2 border p-4')}
    >
      <ThemedTouchableOpacity
        testID={props.testID}
        onPress={onCardPress}
        style={tailwind('mb-2')}
        light={tailwind('border-b-0')}
        dark={tailwind('border-b-0')}
      >
        <View style={tailwind('flex flex-row justify-between mb-4')}>
          <View style={tailwind('flex flex-row flex-1 items-center')}>
            <View style={tailwind('flex flex-col flex-1')}>
              <View style={tailwind('flex flex-row justify-between items-center')}>
                <View style={tailwind('flex-col')}>
                  <View style={tailwind('flex-row items-center')}>
                    <ThemedText
                      light={tailwind('text-gray-400')}
                      dark={tailwind('text-gray-500')}
                      style={tailwind('text-xs mr-1.5')}
                    >
                      {translate('screens/VaultDetailScreen', 'Vault ID')}
                    </ThemedText>
                    <VaultStatusTag status={vaultState.status} testID={`${props.testID}_status`} />
                  </View>
                  <TouchableOpacity
                    style={tailwind('flex flex-row mb-0.5 items-center')}
                    onPress={async () => await openURL(getVaultsUrl(vault.vaultId))}
                  >
                    <ThemedText
                      testID={`${props.testID}_vault_id`}
                      style={tailwind('font-semibold w-56 flex-shrink mr-0.5')}
                      numberOfLines={1}
                      ellipsizeMode='middle'
                    >
                      {vault.vaultId}
                    </ThemedText>
                    <ThemedIcon
                      dark={tailwind('text-darkprimary-500')}
                      iconType='MaterialIcons'
                      light={tailwind('text-primary-500')}
                      name='open-in-new'
                      size={18}
                    />
                  </TouchableOpacity>
                </View>
                {
                  canUseOperations && (
                    <ThemedIcon
                      dark={tailwind('text-gray-200')}
                      iconType='MaterialIcons'
                      light={tailwind('text-black')}
                      name='chevron-right'
                      size={30}
                    />
                  )
                }
              </View>
              <View style={tailwind('flex flex-row')}>
                <ThemedText
                  light={tailwind('text-gray-500')}
                  dark={tailwind('text-gray-300')}
                  style={tailwind('text-xs mr-1')}
                >
                  {translate('components/VaultCard', 'Collaterals:')}
                </ThemedText>
                {
                  (vault.collateralAmounts?.length === 0 || vault.collateralAmounts === undefined) &&
                  (
                    <ThemedText
                      light={tailwind('text-gray-500')}
                      dark={tailwind('text-gray-300')}
                      style={tailwind('text-xs ml-1')}
                      testID={`${props.testID}_collateral_none`}
                    >
                      {translate('components/VaultCard', 'None')}
                    </ThemedText>
                  )
                }
                <TokenIconGroup
                  testID={`${props.testID}_collateral_token_group`}
                  symbols={vault.collateralAmounts?.map(collateral => collateral.displaySymbol)}
                  maxIconToDisplay={3}
                />
              </View>
            </View>
          </View>
        </View>
        {
          ![VaultStatus.Empty, VaultStatus.Ready, VaultStatus.Unknown, VaultStatus.Liquidated].includes(vaultState.status) && (
            <CollateralizationRatioDisplay
              testID={props.testID}
              collateralizationRatio={vault.collateralRatio}
              minCollateralizationRatio={vault.loanScheme.minColRatio}
              totalLoanAmount={vault.loanValue}
              nextCollateralizationRatio={nextCollateralizationRatio?.toFixed(8)}
            />
          )
        }
        <View style={tailwind('flex flex-row flex-wrap -mb-2')}>
          {
            vault.loanAmounts?.length > 0 &&
            (
              <VaultInfo
                label='Active loans' testID={props.testID} tokens={vault.loanAmounts?.map(loan => loan.displaySymbol)}
              />
            )
          }
          <VaultSectionTextRow
            testID={`${props.testID}_total_loan`}
            prefix={VaultStatus.Liquidated === vaultState.status ? '' : '$'}
            value={VaultStatus.Liquidated === vaultState.status ? '-' : getUSDPrecisedPrice(vault.loanValue) ?? '-'}
            lhs={translate('components/VaultCard', 'Total loans (USD)')}
          />
          <VaultSectionTextRow
            testID={`${props.testID}_total_collateral`}
            prefix={VaultStatus.Liquidated === vaultState.status ? '' : '$'}
            value={VaultStatus.Liquidated === vaultState.status ? '-' : getUSDPrecisedPrice(vault.collateralValue)}
            lhs={translate('components/VaultCard', 'Total collateral (USD)')}
          />
        </View>
      </ThemedTouchableOpacity>
      <VaultActionButton vault={vault} canUseOperation={canUseOperations} testID={props.testID} />
    </ThemedView>
  )
}

function VaultActionButton ({
  vault,
  canUseOperation,
  testID
}: { vault: LoanVaultActive, canUseOperation: boolean, testID: string }): JSX.Element | null {
  const navigation = useNavigation<NavigationProp<LoanParamList>>()
  return (
    <ThemedView
      light={tailwind('border-gray-200')}
      dark={tailwind('border-gray-700')}
      style={tailwind('flex flex-row mt-4 items-center flex-wrap -mb-2')}
    >
      {
        new BigNumber(vault.collateralValue).gt(0) && (
          <IconButton
            testID={`${testID}_manage_loans_button`}
            disabled={!canUseOperation || vault.state === LoanVaultState.FROZEN}
            iconLabel={translate('components/VaultCard', 'MANAGE LOANS')}
            style={tailwind('mr-2 mb-2 items-center')}
            onPress={() => {
              navigation.navigate('VaultDetailScreen', {
                vaultId: vault.vaultId,
                tab: TabKey.Loans
              })
            }}
          />
        )
      }
      <IconButton
        testID={`${testID}_edit_collaterals_button`}
        disabled={!canUseOperation}
        iconLabel={translate('components/VaultCard', 'EDIT COLLATERALS')}
        style={tailwind('mr-2 mb-2 items-center')}
        onPress={() => {
          navigation.navigate({
            name: 'EditCollateralScreen',
            params: {
              vaultId: vault.vaultId
            },
            merge: true
          })
        }}
      />
    </ThemedView>
  )
}
