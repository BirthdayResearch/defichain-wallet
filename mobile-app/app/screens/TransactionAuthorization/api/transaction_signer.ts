import { DfTxSigner } from '@store/transaction_queue'
import { WhaleWalletAccount } from '@defichain/whale-api-wallet'
import { CTransactionSegWit } from '@defichain/jellyfish-transaction'
import { Authentication } from '@store/authentication'
import { WalletAlert } from '@components/WalletAlert'
import { translate } from '@translations'
import { INVALID_HASH, MAX_PASSCODE_ATTEMPT } from '@screens/TransactionAuthorization/api/transaction_types'
import { NativeLoggingProps } from '@shared-contexts/NativeLoggingProvider'

export async function execWithAutoRetries (promptPromise: () => Promise<any>, onAutoRetry: (attempts: number) => Promise<void>, retries: number = 0, logger: NativeLoggingProps): Promise<any> {
  try {
    return await promptPromise()
  } catch (e: any) {
    logger.error(e)
    if (e.message === INVALID_HASH && ++retries < MAX_PASSCODE_ATTEMPT) {
      await onAutoRetry(retries)
      return await execWithAutoRetries(promptPromise, onAutoRetry, retries, logger)
    }
    throw e
  }
}

// store/transactionQueue execution
export async function signTransaction (tx: DfTxSigner, account: WhaleWalletAccount, onAutoRetry: (attempts: number) => Promise<void>, retries: number = 0, logger: NativeLoggingProps): Promise<CTransactionSegWit> {
  return await execWithAutoRetries(async () => (await tx.sign(account)), onAutoRetry, retries, logger)
}

// store/authentication execution
export async function authenticateFor<T> (
  promptPassphrase: () => Promise<string>,
  authentication: Authentication<T>,
  onAutoRetry: (attempts: number) => Promise<void>,
  retries: number = 0,
  logger: NativeLoggingProps
): Promise<void> {
  const customJob = async (): Promise<void> => {
    const passphrase = await promptPassphrase()
    const result = await authentication.consume(passphrase)
    return await authentication.onAuthenticated(result)
  }

  return await execWithAutoRetries(customJob, onAutoRetry, retries, logger)
}

export function alertUnlinkWallet (): void {
  WalletAlert({
    title: translate('screens/PinConfirmation', 'Wallet Unlinked'),
    message: translate('screens/PinConfirmation', 'Your wallet was unlinked for your safety due to successive passcode failures. Please use recovery words to restore and set up your wallet again.'),
    buttons: [
      {
        text: translate('screens/PinConfirmation', 'Close'),
        style: 'destructive'
      }
    ]
  })
}
