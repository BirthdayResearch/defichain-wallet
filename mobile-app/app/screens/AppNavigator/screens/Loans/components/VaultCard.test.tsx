
import { VaultCard, VaultCardProps } from './VaultCard'
import { render } from '@testing-library/react-native'
import { LoanVaultState } from '@defichain/whale-api-client/dist/api/loan'

jest.mock('@shared-contexts/ThemeProvider')
jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn()
}))
jest.mock('@shared-contexts/DeFiScanContext')
jest.mock('@gorhom/bottom-sheet', () => ({
  useBottomSheetModal: jest.fn().mockReturnValue({ dismiss: jest.fn() }),
  BottomSheetModal: () => <></>
}))
jest.mock('@components/BottomSheetInfo', () => ({
  BottomSheetInfo: () => <></>
}))

describe('Vault card', () => {
  it('should match snapshot of liquidated vault', async () => {
    const lockedVault: VaultCardProps = {
      vault: {
        vaultId: '22ffasd5ca123123123123123121231061',
        loanScheme: {
          id: '0',
          interestRate: '3',
          minColRatio: '100'
        },
        ownerAddress: 'bcrt1qxzj8pnkeqznvx6xgeepdywus8lkxq3vvmeccyt',
        state: LoanVaultState.IN_LIQUIDATION,
        batches: [],
        batchCount: 0,
        liquidationHeight: 0,
        liquidationPenalty: 0
      },
      testID: 'vault'
    }
    const rendered = render(<VaultCard vault={lockedVault.vault} testID={lockedVault.testID} />)
    expect(rendered.toJSON()).toMatchSnapshot()
  })

  it('should match snapshot of at-risk vault', async () => {
    const atRiskVault: VaultCardProps = {
      vault: {
        vaultId: '22ffasd5ca123123123123123121231061',
        loanAmounts: [],
        collateralRatio: '',
        collateralValue: '',
        collateralAmounts: [],
        loanScheme: {
          id: '0',
          interestRate: '3',
          minColRatio: '100'
        },
        loanValue: '100',
        ownerAddress: 'bcrt1qxzj8pnkeqznvx6xgeepdywus8lkxq3vvmeccyt',
        state: LoanVaultState.MAY_LIQUIDATE,
        informativeRatio: '0',
        interestAmounts: [],
        interestValue: '1'
      },
      testID: 'vault'
    }
    const rendered = render(<VaultCard {...atRiskVault} testID={atRiskVault.testID} />)
    expect(rendered.toJSON()).toMatchSnapshot()
  })

  it('should match snapshot of healthy vault', async () => {
    const safeVault: VaultCardProps = {
      vault: {
        vaultId: '22ffasd5ca123123123123123121231061',
        loanAmounts: [],
        collateralRatio: '',
        collateralValue: '',
        collateralAmounts: [],
        loanScheme: {
          id: '0',
          interestRate: '3',
          minColRatio: '100'
        },
        loanValue: '100',
        ownerAddress: 'bcrt1qxzj8pnkeqznvx6xgeepdywus8lkxq3vvmeccyt',
        state: LoanVaultState.ACTIVE,
        informativeRatio: '0',
        interestAmounts: [],
        interestValue: '1'
      },
      testID: 'vault'
    }
    const rendered = render(<VaultCard vault={safeVault.vault} testID={safeVault.testID} />)
    expect(rendered.toJSON()).toMatchSnapshot()
  })

  it('should match snapshot of active vault', async () => {
    const newVault: VaultCardProps = {
      vault: {
        vaultId: '22ffasd5ca123123123123123121231061',
        loanAmounts: [],
        collateralRatio: '',
        collateralValue: '',
        collateralAmounts: [],
        loanScheme: {
          id: '0',
          interestRate: '3',
          minColRatio: '100'
        },
        loanValue: '100',
        ownerAddress: 'bcrt1qxzj8pnkeqznvx6xgeepdywus8lkxq3vvmeccyt',
        state: LoanVaultState.ACTIVE,
        informativeRatio: '0',
        interestAmounts: [],
        interestValue: '1'
      },
      testID: 'vault'
    }
    const rendered = render(<VaultCard vault={newVault.vault} testID={newVault.testID} />)
    expect(rendered.toJSON()).toMatchSnapshot()
  })
})
