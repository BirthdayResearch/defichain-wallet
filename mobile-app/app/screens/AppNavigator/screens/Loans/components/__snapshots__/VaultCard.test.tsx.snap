// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vault card should match snapshot of active vault 1`] = `
<View
  style={{}}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="vault_"
  >
    <View
      style={{}}
    >
      <View
        style={
          [
            {},
            {},
          ]
        }
        testID="vault"
      >
        <View
          style={{}}
        >
          <View
            style={
              [
                {},
                {
                  "right": 0,
                },
              ]
            }
          />
          <View
            style={{}}
          >
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_max_loan_label"
            >
              Max loan amount
            </Text>
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_max_loan_amount"
            >
              $0.00
            </Text>
          </View>
          <View
            style={{}}
          >
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_total_collateral_label"
            >
              Total collateral
            </Text>
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_total_collateral_amount"
            />
          </View>
        </View>
        <View
          style={{}}
        >
          <react-native-circular-progress-indicator/mock
            activeStrokeColor="#ffffff"
            activeStrokeWidth={3}
            clockwise={false}
            inActiveStrokeColor="#ffffff"
            inActiveStrokeWidth={3}
            maxValue={1}
            radius={58}
            showProgressValue={false}
            value={0.01}
          />
          <View
            style={
              [
                {},
                {
                  "maxHeight": 116,
                },
              ]
            }
          >
            <View
              style={{}}
            >
              <Text
                ellipsizeMode="middle"
                numberOfLines={1}
                style={
                  [
                    {},
                    [
                      [
                        {},
                        {
                          "maxWidth": 124,
                          "minWidth": 10,
                        },
                      ],
                      {},
                    ],
                  ]
                }
                testID="vault_vault_id"
              >
                22ffasd5ca123123123123123121231061
              </Text>
              <Text
                style={
                  [
                    {},
                    [
                      {},
                      {},
                    ],
                  ]
                }
                testID="vault_min_ratio"
              >
                0.00%
              </Text>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  [
                    {},
                    [
                      {},
                      {},
                    ],
                  ]
                }
                testID="vault_borrow_button"
              >
                Borrow
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Vault card should match snapshot of at-risk vault 1`] = `
<View
  style={{}}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="vault_"
  >
    <View
      style={{}}
    >
      <View
        style={
          [
            {},
            {},
          ]
        }
        testID="vault"
      >
        <View
          style={{}}
        >
          <View
            style={
              [
                {},
                {
                  "right": 0,
                },
              ]
            }
          />
          <View
            style={{}}
          >
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_max_loan_label"
            >
              Max loan amount
            </Text>
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_max_loan_amount"
            >
              $0.00
            </Text>
          </View>
          <View
            style={{}}
          >
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_total_collateral_label"
            >
              Total collateral
            </Text>
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_total_collateral_amount"
            />
          </View>
        </View>
        <View
          style={{}}
        >
          <react-native-circular-progress-indicator/mock
            activeStrokeColor="#ffffff"
            activeStrokeWidth={3}
            clockwise={false}
            inActiveStrokeColor="#ffffff"
            inActiveStrokeWidth={3}
            maxValue={1}
            radius={58}
            showProgressValue={false}
            value={0.01}
          />
          <View
            style={
              [
                {},
                {
                  "maxHeight": 116,
                },
              ]
            }
          >
            <View
              style={{}}
            >
              <Text
                ellipsizeMode="middle"
                numberOfLines={1}
                style={
                  [
                    {},
                    [
                      [
                        {},
                        {
                          "maxWidth": 124,
                          "minWidth": 10,
                        },
                      ],
                      {},
                    ],
                  ]
                }
                testID="vault_vault_id"
              >
                22ffasd5ca123123123123123121231061
              </Text>
              <Text
                style={
                  [
                    {},
                    [
                      {},
                      {},
                    ],
                  ]
                }
                testID="vault_min_ratio"
              >
                0.00%
              </Text>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  [
                    {},
                    [
                      {},
                      {},
                    ],
                  ]
                }
                testID="vault_borrow_button"
              >
                Borrow
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Vault card should match snapshot of healthy vault 1`] = `
<View
  style={{}}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="vault_"
  >
    <View
      style={{}}
    >
      <View
        style={
          [
            {},
            {},
          ]
        }
        testID="vault"
      >
        <View
          style={{}}
        >
          <View
            style={
              [
                {},
                {
                  "right": 0,
                },
              ]
            }
          />
          <View
            style={{}}
          >
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_max_loan_label"
            >
              Max loan amount
            </Text>
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_max_loan_amount"
            >
              $0.00
            </Text>
          </View>
          <View
            style={{}}
          >
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_total_collateral_label"
            >
              Total collateral
            </Text>
            <Text
              style={
                [
                  {},
                  [
                    {},
                    {},
                  ],
                ]
              }
              testID="vault_total_collateral_amount"
            />
          </View>
        </View>
        <View
          style={{}}
        >
          <react-native-circular-progress-indicator/mock
            activeStrokeColor="#ffffff"
            activeStrokeWidth={3}
            clockwise={false}
            inActiveStrokeColor="#ffffff"
            inActiveStrokeWidth={3}
            maxValue={1}
            radius={58}
            showProgressValue={false}
            value={0.01}
          />
          <View
            style={
              [
                {},
                {
                  "maxHeight": 116,
                },
              ]
            }
          >
            <View
              style={{}}
            >
              <Text
                ellipsizeMode="middle"
                numberOfLines={1}
                style={
                  [
                    {},
                    [
                      [
                        {},
                        {
                          "maxWidth": 124,
                          "minWidth": 10,
                        },
                      ],
                      {},
                    ],
                  ]
                }
                testID="vault_vault_id"
              >
                22ffasd5ca123123123123123121231061
              </Text>
              <Text
                style={
                  [
                    {},
                    [
                      {},
                      {},
                    ],
                  ]
                }
                testID="vault_min_ratio"
              >
                0.00%
              </Text>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  [
                    {},
                    [
                      {},
                      {},
                    ],
                  ]
                }
                testID="vault_borrow_button"
              >
                Borrow
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Vault card should match snapshot of liquidated vault 1`] = `
<View
  style={{}}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="vault_IN LIQUIDATION"
  >
    <View
      style={
        [
          {},
          {},
        ]
      }
      testID="vault_IN LIQUIDATION_vault"
    >
      <View
        style={{}}
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            {
              "height": 64,
              "width": 74,
            }
          }
          testID="vault_IN LIQUIDATION_liquidated_vault_image"
        />
        <View
          style={{}}
        >
          <View
            accessibilityState={
              {
                "disabled": false,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <Text
              ellipsizeMode="middle"
              numberOfLines={1}
              style={
                [
                  {},
                  [
                    [
                      {},
                      {
                        "maxWidth": 124,
                        "minWidth": 10,
                      },
                    ],
                    {},
                  ],
                ]
              }
              testID="vault_IN LIQUIDATION_vault_id"
            >
              22ffasd5ca123123123123123121231061
            </Text>
            <Text />
          </View>
          <Text
            style={
              [
                {},
                [
                  {},
                  {},
                ],
              ]
            }
            testID="vault_IN LIQUIDATION_vault_description"
          >
            Liquidated vault
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
